src/Features/CSharp/Portable/CodeRefactorings/PullMemberUp/CSharpPullMemberUpCodeRefactoringProvider.cs

```diff
@@ -25,42 +28,11 @@ public CSharpPullMemberUpCodeRefactoringProvider() : base(null)
         {
         }
 
-        protected override bool IsSelectionValid(TextSpan span, SyntaxNode selectedNode)
+        protected override async Task<SyntaxNode> GetSelectedNodeAsync(Document document, TextSpan span, CancellationToken cancellationToken)
         {
-            var identifier = GetIdentifier(selectedNode);
-            if (identifier == default)
-            {
-                return false;
-            }
-            else if (identifier.FullSpan.Contains(span) && span.Contains(identifier.Span))
-            {
-                // Selection lies within the identifier's span
-                return true;
-            }
-            else if (identifier.Span.Contains(span) && span.IsEmpty)
-            {
-                // Cursor stands on the identifier
-                return true;
-            }
-            else
-            {
-                return false;
-            }
-        }
-
-        private SyntaxToken GetIdentifier(SyntaxNode selectedNode)
-        {
-            switch (selectedNode)
-            {
-                case MemberDeclarationSyntax memberDeclarationSyntax:
-                    // Nested type is checked in before this method is called.
-                    return memberDeclarationSyntax.GetNameToken();
-                case VariableDeclaratorSyntax variableDeclaratorSyntax:
-                    // It handles multiple fields or events declared in one line
-                    return variableDeclaratorSyntax.Identifier;
-                default:
-                    return default;
-            }
+            var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>();
+            return await refactoringHelperService.TryGetSelectedNodeAsync(
+                document, span, n => n is MemberDeclarationSyntax || n is VariableDeclaratorSyntax, cancellationToken).ConfigureAwait(false);
```

Comment:
woudl it be possible to instead write this as:

```c#
var memberDecl = await refactoringHelperService.TryGetSelectedNodeAsync<MemberDeclarationSyntax>(document, span, cancellationToken);
var varDecl = await refactoringHelperService.TryGetSelectedNodeAsync<VariableDeclaratorSyntax>(document, span, cancellationToken);

return memberDecl ?? varDecl;
```

?

Honestly, it reads a lot better to me to just have the caller say "here's the type of node i want" and that's the entirety of the surface area of this call.

Reply:
I mean sure, but that means running everything twice and even though this isn't hot path it still feels wasteful for me. And using lambda for type test doesn't feel all that bad or complex.

Reply:
this is not at all a hotpath :)  And i thnk it's really nice to keep things conceptually clear.  I really liek the "just let me know if this node-was conceptually selected by the user".  

premature optimization yadda yadda.

Basically, let's start simple and clear.  And if there is evidence or information to make us feel like the extra complexity is warranted, we go that route.  (in other words, the Go philosophy :D).

Reply:
Ok, while I'm not sure I completely agree with that I see the idea. Let's switch it to two simple calls & try to stick to simple calls whenever possible.